src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -356,45 +363,51 @@ void CDROM_Interface_Ioctl::CdAudioCallback(const uint16_t requested_frames)
 		cdrom_read_audio cd = {};
 		cd.addr.lba         = current_sector;
 		cd.addr_format      = CDROM_LBA;
-		cd.buf = reinterpret_cast<uint8_t*>(audio_buffer.data());
+		cd.buf = reinterpret_cast<uint8_t*>(input_buffer.data());
 
 		// Maximum this ioctl can read in a single call is 1 second (75
 		// redbook frames)
 		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
-		constexpr int PcmFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
-		                                         BYTES_PER_REDBOOK_PCM_FRAME;
-		frames_in_buffer = cd.nframes * PcmFramesPerRedbookFrame;
+		constexpr auto PcmSamplesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                           REDBOOK_BPS;
+		size_t samples = cd.nframes * PcmSamplesPerRedbookFrame;
 		current_sector += cd.nframes;
 		sectors_remaining -= cd.nframes;
+
 		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
 			// Write out silence if something goes wrong
-			std::fill(audio_buffer.begin(), audio_buffer.end(), 0);
+			std::fill(input_buffer.begin(), input_buffer.end(), 0);
 
 #ifdef DEBUG_IOCTL
 			LOG_WARNING("CDROM_IOCTL: CdAudioCallback: CDROMREADAUDIO ioctl failed");
 #endif
 		}
+
+		output_buffer.clear();
+		for (size_t i = 0; i < samples; i += 2) {
+			const int16_t left  = host_to_le16(input_buffer[i]);
+			const int16_t right = host_to_le16(input_buffer[i + 1]);
+
+			// output_buffer.emplace_back(left, right);
+
+			// For some reason, emplace_back is much slower
+			// in debug build and causes noticible popping
+			// This makes the loop 4ms instead of 10ms on my system.
+			AudioFrame frame(left, right);
```

Comment:
It's fine, please use emplace and remove this comment. This happens because filling up 1s worth of data takes just a bit too long, which results on starving the audio buffer occasionally in debug builds. You can increase the buffer size in your config to work around this somewhat in debug builds, but it's normal. Debug builds are dog slow with certain settings and can crackle like hell... Also, probably you won't get crackling if you used a standalone audio player program with low cycles settings. Quake with max cycles setting will cause crackles in debug builds, it's not a surprise. Maybe try Monkey Island CD edition with around 3000-5000 cycles. But ultimately it's ok. We could put comments like these into every single file about debug builds 😄

Reply:
I've done the on demand float conversations and that seems to have solved the performance issues on debug builds.